Implement sensei-lms/course-theme-complete-lesson block.

dadish opened the pull request:
Fixes #4405 

### Changes proposed in this Pull Request

* Implements `sensei-lms/course-theme-complete-lesson` block.

### Testing instructions

<!--
If the functionality of this PR is behind a feature flag, please include the
following testing step, using the correct name for the feature flag. (If you
aren't sure, just ignore this step)

* Enable the feature flag: `add_filter( 'sensei_feature_flag_{THE_FLAG_NAME}', '__return_true' );`
-->
* Enable the feature flag: `add_filter( 'sensei_feature_flag_course-theme', '__return_true' );`
* Confirm the "Mark complete" button is rendered as **primary** CTA for a lesson.
* Confirm the "Mark complete" button is rendered as **secondary** CTA for a lesson with quiz but not requiring to pass.
* Confirm the "Mark complete" button is rendered as **disabled** CTA for a lesson with a pre-requisite.
* Confirm the "Mark complete" button **is not** rendered if lesson has quiz that requires passing.
* Confirm the "Mark complete" button **is not** rendered if lesson is already completed by the user.
* Confirm the "Mark complete" button completes the lesson and navigates to the next one when clicked.

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video


https://user-images.githubusercontent.com/2578542/143580260-9234069f-7092-4b5a-b26b-6701da0d2f4e.mp4

dadish commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Complete Lesson block](https://trello.com/c/aOMAPGDr/1224-complete-lesson-block)

renatho commented:
@dadish, I was checking the Figma after the last design changes, the "Completed" doesn't exist anymore. Also, a new variation is with the button disabled if it contains a pre-requisite and the user can't take the lesson.

dadish commented:
> @dadish, I was checking the Figma after the last design changes, the "Completed" doesn't exist anymore. Also, a new variation is with the button disabled if it contains a pre-requisite and the user can't take the lesson.

@renatho
- Changed the "Completed" state, now it renders empty string if lesson is already complete.
- Added "Disabled" state that is rendered when lesson has a pre-requisite.
- Updated styles and tests.

renatho commented on includes/blocks/course-theme/class-complete-lesson.php:
```diff
@@ -0,0 +1,95 @@
+		$text      = esc_html( __( 'Complete lesson', 'sensei-lms' ) );
+
+		return ( "
+			<form class='sensei-course-theme-complete-lesson-form' method='POST' action='{$permalink}'>
```
Here I have that same observation about the single quotes (since we use double quotes for the HTML in all the other parts, it would be nice to keep the same standard), but we can change that when moving it to the actions block! =)

dadish replied:
Updated! 👍 

Btw, what would be the reason we're following this convention? Why HTML attribute values should be double quoted instead of single quotes?

alexsanford replied:
FWIW here's the [WP Coding Standards section](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/#quotes) on quotes for HTML attributes. According to the WP Standard, quotes are necessary, but single vs double doesn't matter.

renatho replied:
> what would be the reason we're following this convention?

Honestly, I don't have the answer to why we're following double quotes as rule in the plugin. I think just because it was the preference of the team that started writing the plugin. :)

dadish replied:
> FWIW here's the WP Coding Standards section on quotes for HTML attributes. According to the WP Standard, quotes are necessary, but single vs double doesn't matter.

> Honestly, I don't have the answer to why we're following double quotes as rule in the plugin. I think just because it was the preference of the team that started writing the plugin. :)

Well in that case, moving forward I think it should be OK to wrap our html with double quotes and use single quotes for attributes. It's much better to write and read that way.

#### Double quote HTML attributes

```php
$html = '
  <form class="sensei-course-theme-complete-lesson-form" method="POST" action="' . $permalink . '">
    ' . $nonce . '
    <input type="hidden" name="quiz_action" value="lesson-complete" />
    <button type="submit" class="sensei-course-theme-complete-lesson ' . $secondary . '" ' . $disabled . '>
      ' . $text . '
    </button>
  </form>
';
```

#### Single quote HTML attributes

```php
$html = "
  <form class='sensei-course-theme-complete-lesson-form' method='POST' action='{$permalink}'>
    {$nonce}
    <input type='hidden' name='quiz_action' value='lesson-complete' />
    <button type='submit' class='sensei-course-theme-complete-lesson{$secondary}' {$disabled}>
      {$text}
    </button>
  </form>
";
```

Declarative FTW!

renatho replied:
That would be okay for me too! I think it's just a question of setting a convention for our codebase. =) Maybe it would be a topic to open a discussion on a P2 and define?

Also, one thing I thought a while ago was to create templates for any HTML we write in PHP (maybe with the templates we have currently - that the user can override, or a template lib just to separate the markup). So we would avoid writing HTML inside PHP.

renatho reviewed (approved):
Great job! And great test coverage too! =)

renatho reviewed (approved):
🚀
